### PR TITLE
Add unit tests for BlockMasterClient

### PR DESCRIPTION
**Pull request by YangchenYe323**
### What changes are proposed in this pull request?
add more unit tests in BlockMasterClientTest that tests for expected grpc behaviors.

### Why are the changes needed?
improve unit-test coverage of BlockMasterClient (20% -> 90%)

### Does this PR introduce any user facing changes?
No

### CODE REVIEW

**Comment by YangchenYe323** [comment]
@dbw9580 May you take a look at my PR? Thanks!

File: core/server/worker/pom.xml [hunk 1]
```diff
@@ -109,7 +109,10 @@
       <artifactId>hamcrest</artifactId>
       <scope>test</scope>
     </dependency>
-
```
Comment (Xenorith) [inline core/server/worker/pom.xml:112]: keep a new line between the last external test dependency and the internal test dependency comment

File: core/server/worker/pom.xml [hunk 2]
```diff
@@ -109,7 +109,10 @@
...
+    <dependency>
+      <groupId>io.grpc</groupId>
+      <artifactId>grpc-testing</artifactId>
```
Comment (Xenorith) [inline core/server/worker/pom.xml:114]: if this is supposed to be only for testing, add the test scope similar to the dependency above?

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 3]
```diff
@@ -77,4 +121,314 @@ public void convertBlockListMapToProtoMergeDirsInSameTier() {
               new HashSet<>(ssdProtoBlockList));
     }
   }
+
+  @Test
+  public void testCommitBlock() throws Exception {
+    HashMap<Long, Long> committedBlocks = new HashMap<>();
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:127]: Use a ConcurrentHashMap for thread safety?

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 4]
```diff
@@ -77,4 +121,314 @@ public void convertBlockListMapToProtoMergeDirsInSameTier() {
...
+            responseObserver.onCompleted();
+          }
+        });
+
+    BlockMasterClient client = new MockStubBlockMasterClient(
+        MasterClientContext.newBuilder(ClientContext.create(mConf)).build(),
+        mockChannel
+    );
+
+    assertEquals(CommandType.Nothing, client.heartbeat(
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:255]: Also need to verify that the server receives the same capacity and metrics info as the client has sent.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 5]
```diff
@@ -77,4 +121,314 @@ public void convertBlockListMapToProtoMergeDirsInSameTier() {
...
+            .builder()
+            .withMaxDuration(Duration.of(500, ChronoUnit.MILLIS))
+            .withInitialSleep(Duration.of(0, ChronoUnit.MILLIS))
+            .withMaxSleep(Duration.of(100, ChronoUnit.MILLIS))
+            .withSkipInitialSleep()
+            .build());
+  }
+
+  public void register(boolean stream) throws Exception {
+    final long workerId = 0L;
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:316]: Use a non zero value here, as the default value for the worker id in `RegisterWorkerPRequest` is 0, if that field is not set. Otherwise, should the test subject forget to set the correct worker ID in the request and leave it to default, this test would fail to catch that.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 6]
```diff
@@ -77,4 +121,314 @@ public void convertBlockListMapToProtoMergeDirsInSameTier() {
...
+          storageTierAliases,
+          totalBytesOnTiers,
+          usedBytesOnTiers,
+          currentBlocksOnLocation,
+          lostStorage,
+          configList);
+    }
+
+    assertEquals(1, registeredWorkerIds.size());
+    assertEquals(workerId, (long) registeredWorkerIds.get(0));
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:385]: Verify other arguments passed to this method, e.g. `currentBlocksOnLocation` is the same.

**Review by dbw9580 — CHANGES_REQUESTED** [review]
First batch of comments, thanks for the contribution!

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 7]
```diff
@@ -30,6 +68,12 @@
 import java.util.stream.Collectors;
 
 public class BlockMasterClientTest {
+
+  @Rule
+  public final GrpcCleanupRule mGrpcCleanup = new GrpcCleanupRule();
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:73]: https://github.com/grpc/grpc-java/blob/c48b01a46c31cf45e8c4d66e8afc4e46dc7e99cc/testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java#L74

I wonder why are good things always experimental?

File: core/server/worker/pom.xml (hunk 2 above)
Comment (YangchenYe323) [reply core/server/worker/pom.xml:114]: fixed

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 4 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:291]: Makes sense. I made some verification in the new commit. Does that looks good to you?

File: core/server/worker/pom.xml (hunk 1 above)
Comment (YangchenYe323) [reply core/server/worker/pom.xml:112]: Fixed

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 8]
```diff
@@ -77,4 +121,314 @@ public void convertBlockListMapToProtoMergeDirsInSameTier() {
...
+            delegatesTo(delegate));
+
+    // create a mock server using mock service
+    String serverName = InProcessServerBuilder.generateName();
+    mGrpcCleanup.register(InProcessServerBuilder
+        .forName(serverName).directExecutor().addService(mockService).build().start());
+
+    // return the channel to the mock server
+    return mGrpcCleanup.register(
+        InProcessChannelBuilder.forName(serverName).directExecutor().build());
```
[the remark this answers is not in the document]
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java]: Fixed in new commit. Experimental APIs are replaced

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 9]
```diff
@@ -77,4 +121,314 @@ public void convertBlockListMapToProtoMergeDirsInSameTier() {
...
+        workerId,
+        capacityBytesOnTiers,
+        usedBytesOnTiers,
+        removedBlocks,
+        addedBlocks,
+        lostStorage,
+        metrics).getCommandType());
+  }
+
+  @Test(expected = FailedToAcquireRegisterLeaseException.class)
```
[the remark this answers is not in the document]
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java]: Fixed

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 5 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java]: Got it. Fixed

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 10]
```diff
@@ -77,4 +121,314 @@ public void convertBlockListMapToProtoMergeDirsInSameTier() {
...
+    client.acquireRegisterLeaseWithBackoff(
+        0L,
+        1,
+        ExponentialTimeBoundedRetry
+            .builder()
+            .withMaxDuration(Duration.of(500, ChronoUnit.MILLIS))
+            .withInitialSleep(Duration.of(0, ChronoUnit.MILLIS))
+            .withMaxSleep(Duration.of(100, ChronoUnit.MILLIS))
+            .withSkipInitialSleep()
+            .build());
```
[the remark this answers is not in the document]
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java]: Fixed

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 6 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:468]: Fixed in new commit

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 11]
```diff
@@ -80,23 +79,26 @@
...
-          PropertyKey.MASTER_RPC_PORT, 9999,
-          PropertyKey.MASTER_RPC_ADDRESSES, ImmutableList.of(TEST_SOCKET_ADDRESS),
+      .of(
+          PropertyKey.MASTER_RPC_ADDRESSES, ImmutableList.of(TEST_SOCKET_ADDRESS_STRING),
           PropertyKey.USER_RPC_RETRY_MAX_DURATION, "5s",
           PropertyKey.USER_RPC_RETRY_BASE_SLEEP_MS, "100ms",
           PropertyKey.USER_RPC_RETRY_MAX_SLEEP_MS, "500ms",
-          PropertyKey.SECURITY_AUTHENTICATION_TYPE, AuthType.NOSASL),
+          PropertyKey.SECURITY_AUTHENTICATION_TYPE, AuthType.NOSASL,
+          PropertyKey.USER_CONF_CLUSTER_DEFAULT_ENABLED, false),
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:97]: Add a comment here that explains why cluster defaults cannot be acquired.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 12]
```diff
@@ -80,23 +79,26 @@
 
 public class BlockMasterClientTest {
 
+  // alert: you CANNOT just set TEST_SOCKET_ADDRESS to
+  // configuration, because it will get turned to string "localhost/127.0.0.1:9999"
+  // This string cannot be parsed accurately by the current parsing algorithm
   private static final InetSocketAddress TEST_SOCKET_ADDRESS =
-      new InetSocketAddress("0.0.0.0", 9999);
+      new InetSocketAddress("localhost", 9999);
+  private static final String TEST_SOCKET_ADDRESS_STRING = "localhost:9999";
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:87]: nit: does `TEST_SOCKET_ADDRESS.toString()` work here? If so, we can get rid of this additional constant.

**Review by dbw9580 — APPROVED** [review]
LGTM with a few minor comments

File: core/server/worker/pom.xml [hunk 13]
```diff
@@ -109,6 +109,11 @@
       <artifactId>hamcrest</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>io.grpc</groupId>
+      <artifactId>grpc-testing</artifactId>
```
Comment (dbw9580) [inline core/server/worker/pom.xml:114]: Is this dependency still used after you removed the `GrpcCleanupRule`, etc. ?

**Comment by dbw9580** [comment]
@jiacheliu3 do you mind taking a look at this unit test?

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 12 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:87]: It doesn't work here for the reason stated in the comment🤦‍♂️. The `InetSocketAddress` will be turned into "localhost/127.0.0.1:9999". This string cannot be parsed correctly by the current algorithm: [NetworkAddressUtils#parseInetSocketAddress](https://github.com/Alluxio/alluxio/blob/master/core/common/src/main/java/alluxio/util/network/NetworkAddressUtils.java).

Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:87]: I tried some alternatives like `new InetSocketAddress("127.0.0.1", 9999)` and `new InetSocketAddress("0.0.0.0", 9999)`. They don't seem to produce connectable addresses because the constructor expects a host name in that argument.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 11 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:100]: Done

File: core/server/worker/pom.xml (hunk 13 above)
Comment (YangchenYe323) [reply core/server/worker/pom.xml]: It's not used. Removed in the new commit.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 12 above)
Comment (jiacheliu3) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:87]: wow interesting, do you think [NetworkAddressUtils#parseInetSocketAddress](https://github.com/Alluxio/alluxio/blob/master/core/common/src/main/java/alluxio/util/network/NetworkAddressUtils.java) can be fixed easily? If not, we can leave it be.

I'm fine with keeping all necessary variables here. It's just the comment seems a bit verbose and distractive.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 14]
```diff
@@ -12,24 +12,109 @@
...
+          PropertyKey.USER_CONF_CLUSTER_DEFAULT_ENABLED, false),
+      Configuration.modifiableGlobal());
+
+  // mock rpc server managed by test suite
+  private GrpcServer mServer;
+
+  private final AlluxioConfiguration mConf = Configuration.global();
+
+  @Test
+  public void testClientInfo() {
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:109]: btw you don't need the string "test" in the test method name, they are marked as `@Test` already. So this one can be just named `clientInfo()`. This applies to all other test methods.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 15]
```diff
@@ -77,4 +162,343 @@ public void convertBlockListMapToProtoMergeDirsInSameTier() {
...
+  // create a mock grpc server that uses delegate to handle rpc calls
+  private void createMockService(
+      BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase delegate) throws IOException {
+
+    // make sure to clean up resources
+    cleanUp();
+
+    // create mock service handler
+    BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase mockService =
+        mock(
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:490]: For the same reason mentioned in the other PR, is it possible to replace this mock with a real child class?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Added some general comments. Unfortunately I don't have enough bandwidth to scrutinize the test cases themselves and I will rely on Bowen's judgement. I will keep an eye out on the general coding style and help. Thanks!

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 12 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:87]: > wow interesting, do you think [NetworkAddressUtils#parseInetSocketAddress](https://github.com/Alluxio/alluxio/blob/master/core/common/src/main/java/alluxio/util/network/NetworkAddressUtils.java) can be fixed easily? If not, we can leave it be.
> 
> I'm fine with keeping all necessary variables here. It's just the comment seems a bit verbose and distractive.

I would say yes. We could fix it by splitting the hostname by "/" and take the first part.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 15 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:490]: Mock is redundant here, we can just use the anonymous `BlockMasterWorkerServiceImplBase` directly. I'm not sure if there would be benefit of creating an explicit child class, since different test methods overrides different methods and there are no obvious common functionalities.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 14 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java]: Fixed

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 12 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:85]: Since "/" is not a valid character in a hostname, this is not likely to break present working hostnames.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java [hunk 16]
```diff
@@ -328,27 +326,27 @@ public void blockHeartbeat(BlockHeartbeatPRequest request,
...
 
   @Test
-  public void testRegister() throws Exception {
+  public void registerWithoutStream() throws Exception {
     register(false);
   }
 
   @Test
-  public void testRegisterStream() throws Exception {
+  public void testRegisterWithStream() throws Exception {
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java:345]: nit: `registerWithStream`

File: core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java (hunk 15 above)
Comment (jiacheliu3) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockMasterClientTest.java]: Yes we totally can. Your current approach is good.

**Review by jiacheliu3 — APPROVED** [review]
LGTM with a minor comment. Nice work!

**Comment by jiacheliu3** [comment]
The current unit tests are more about positive test cases, which are the fundamental guarantee like "Here are my APIs and this is how you are expected to use them". If you are to add more tests on them, you can add negative test cases where you verify what happens on exceptions and write down the assertions(expected behaviors) in the form of unit tests. That is beyond the scope of one PR, feel free to do that in a separate PR, now that we have the basics covered in this PR and feel more secure about the existing API.

**Comment by dbw9580** [comment]
@jiacheliu3 can we merge this?

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please
